File: android/src/main/kotlin/com/g12/flutter_native_admob_ads/FlutterNativeAdmobAdsPlugin.kt
package com.g12.flutter_native_admob_ads

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.LinearLayout
import com.google.android.gms.ads.*
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdOptions
import com.google.android.gms.ads.nativead.NativeAdView
import android.os.Bundle
import com.google.ads.mediation.admob.AdMobAdapter
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.*
import kotlin.collections.HashMap

/** FlutterNativeAdmobAdsPlugin */
class FlutterNativeAdmobAdsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var context: Context? = null
    private var activity: Activity? = null
    private val nativeAds = HashMap<String, NativeAd>()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_native_admob_ads")
        channel.setMethodCallHandler(this)
        
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "flutter_native_ad_view", 
            FlutterNativeAdViewFactory(nativeAds)
        )
        
        MobileAds.initialize(context!!) { }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "loadNativeAd" -> {
                loadNativeAd(call, result)
            }
            "disposeNativeAd" -> {
                disposeNativeAd(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun loadNativeAd(call: MethodCall, result: Result) {
        val adId = if (call.argument<Boolean>("isTesting") == true) {
            "ca-app-pub-3940256099942544/2247696110"
        } else {
            call.argument<String>("adId")
        }

        if (adId == null) {
            result.error("MISSING_AD_ID", "Ad ID is required", null)
            return
        }

        val adsCount = call.argument<Int>("adsCount") ?: 1
        val adRequestMap = call.argument<Map<String, Any?>>("adRequest")
        val adsToReturn = Collections.synchronizedList(mutableListOf<Map<String, Any?>>())
        
        loadAdsSequentially(adId, adsCount, adsToReturn, adRequestMap, result)
    }

    private fun loadAdsSequentially(
        adId: String, 
        remainingCount: Int, 
        loadedAds: MutableList<Map<String, Any?>>, 
        adRequestMap: Map<String, Any?>?,
        result: Result
    ) {
        if (remainingCount <= 0) {
            activity?.runOnUiThread {
                if (loadedAds.isNotEmpty()) {
                    result.success(loadedAds)
                } else {
                    result.error("LOAD_FAILED", "No ads loaded", null)
                }
            }
            return
        }

        val adLoader = AdLoader.Builder(context!!, adId)
            .forNativeAd { nativeAd ->
                val adMap = mapNativeAd(nativeAd)
                loadedAds.add(adMap)
                
                // Trigger next load
                loadAdsSequentially(adId, remainingCount - 1, loadedAds, adRequestMap, result)
            }
            .withAdListener(object : AdListener() {
                var currentAdId: String? = null

                private fun ensureAdId(): String? {
                    if (currentAdId == null && loadedAds.isNotEmpty()) {
                        currentAdId = loadedAds.last()["id"] as String?
                    }
                    return currentAdId
                }

                override fun onAdFailedToLoad(adError: LoadAdError) {
                    loadAdsSequentially(adId, remainingCount - 1, loadedAds, adRequestMap, result)
                }

                override fun onAdOpened() {
                    ensureAdId()?.let { id ->
                        channel.invokeMethod("onAdOpened", id)
                    }
                }

                override fun onAdClicked() {
                    ensureAdId()?.let { id ->
                        channel.invokeMethod("onAdClicked", id)
                    }
                }

                override fun onAdClosed() {
                    ensureAdId()?.let { id ->
                        channel.invokeMethod("onAdClosed", id)
                    }
                }

                override fun onAdImpression() {
                    ensureAdId()?.let { id ->
                        channel.invokeMethod("onAdImpression", id)
                    }
                }
            })
            .withNativeAdOptions(NativeAdOptions.Builder().build())
            .build()

        val adRequest = buildAdRequest(adRequestMap)
        adLoader.loadAd(adRequest)
    }

    private fun buildAdRequest(map: Map<String, Any?>?): AdRequest {
        val builder = AdRequest.Builder()
        if (map == null) return builder.build()

        val keywords = map["keywords"] as? List<String>
        keywords?.forEach { builder.addKeyword(it) }

        val contentUrl = map["contentUrl"] as? String
        contentUrl?.let { builder.setContentUrl(it) }

        val neighboringContentUrls = map["neighboringContentUrls"] as? List<String>
        neighboringContentUrls?.let { builder.setNeighboringContentUrls(it) }

        val httpTimeoutMillis = map["httpTimeoutMillis"] as? Int
        httpTimeoutMillis?.let { builder.setHttpTimeoutMillis(it) }

        // mediationExtrasIdentifier is not supported by AdRequest.Builder

        // Extras and Non-personalized ads
        val extras = Bundle()
        val customExtras = map["extras"] as? Map<String, String>
        customExtras?.forEach { (key, value) -> extras.putString(key, value) }

        val nonPersonalizedAds = map["nonPersonalizedAds"] as? Boolean
        if (nonPersonalizedAds == true) {
            extras.putString("npa", "1")
        }

        if (!extras.isEmpty) {
            builder.addNetworkExtrasBundle(AdMobAdapter::class.java, extras)
        }

        // Other Mediation Extras
        val mediationExtrasList = map["mediationExtras"] as? List<Map<String, Any?>>
        mediationExtrasList?.forEach { extraMap ->
            val adapterClassName = extraMap["androidClassName"] as? String
            val innerExtras = extraMap["extras"] as? Map<String, Any?>
            if (adapterClassName != null && innerExtras != null) {
                try {
                    // Use addNetworkExtrasBundle(Class<? extends MediationExtrasReceiver>, Bundle)
                    val bundle = Bundle()
                    innerExtras.forEach { (k, v) ->
                        when (v) {
                            is String -> bundle.putString(k, v)
                            is Int -> bundle.putInt(k, v)
                            is Boolean -> bundle.putBoolean(k, v)
                            is Double -> bundle.putDouble(k, v)
                        }
                    }
                    val receiverClass = Class.forName(adapterClassName) as Class<out com.google.android.gms.ads.mediation.MediationExtrasReceiver>
                    builder.addNetworkExtrasBundle(receiverClass, bundle)
                } catch (e: Exception) {
                    // Ignore or log error
                }
            }
        }

        return builder.build()
    }

    private fun mapNativeAd(nativeAd: NativeAd): Map<String, Any?> {
        val id = UUID.randomUUID().toString()
        val map = HashMap<String, Any?>()
        
        map["id"] = id
        map["headline"] = nativeAd.headline
        map["body"] = nativeAd.body
        map["advertiser"] = nativeAd.advertiser
        map["cta"] = nativeAd.callToAction
        map["starRating"] = nativeAd.starRating
        map["store"] = nativeAd.store
        map["price"] = nativeAd.price
        
        // Icon
        nativeAd.icon?.let { adIcon ->
            val iconMap = HashMap<String, Any?>()
            iconMap["url"] = adIcon.uri?.toString()
            adIcon.drawable?.let {
                iconMap["width"] = it.intrinsicWidth.toDouble()
                iconMap["height"] = it.intrinsicHeight.toDouble()
            }
            iconMap["scale"] = adIcon.scale
            map["icon"] = iconMap
        }
        
        // Images
        val images = nativeAd.images
        if (images.isNotEmpty()) {
            val imageListData = images.map { adImage ->
                val imageMap = HashMap<String, Any?>()
                imageMap["url"] = adImage.uri?.toString()
                adImage.drawable?.let {
                    imageMap["width"] = it.intrinsicWidth.toDouble()
                    imageMap["height"] = it.intrinsicHeight.toDouble()
                }
                imageMap["scale"] = adImage.scale
                imageMap
            }
            map["images"] = imageListData
            map["cover"] = imageListData.firstOrNull()?.get("url")
        } else {
            map["images"] = null
            map["cover"] = null
        }
        
        // Media Content (Aspect Ratio)
        nativeAd.mediaContent?.let { media ->
            map["aspectRatio"] = media.aspectRatio.toDouble()
        }
        
        val adChoices = nativeAd.adChoicesInfo
        map["adChoicesUrl"] = "https://adssettings.google.com/whythisad"
        map["adChoicesText"] = adChoices?.text?.toString()

        synchronized(nativeAds) {
            nativeAds[id] = nativeAd
        }

        return map
    }


    private fun disposeNativeAd(call: MethodCall, result: Result) {
        val id = call.argument<String>("id")
        synchronized(nativeAds) {
            nativeAds.remove(id)
        }
        result.success(null)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: ios/Classes/FlutterNativeAdmobAdsPlugin.swift
import Flutter
import UIKit
import GoogleMobileAds

public class FlutterNativeAdmobAdsPlugin: NSObject, FlutterPlugin, GADAdLoaderDelegate, GADNativeAdLoaderDelegate {
  private var adLoader: GADAdLoader?
  internal var nativeAds = [String: GADNativeAd]()
  private var loadResults = [String: (loaded: [[String: Any?]], failed: Int, total: Int, result: FlutterResult)]()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_native_admob_ads", binaryMessenger: registrar.messenger())
    let instance = FlutterNativeAdmobAdsPlugin()
    instance.channelInstance = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
    
    registrar.register(
      FlutterNativeAdPlatformViewFactory(plugin: instance),
      withId: "flutter_native_ad_view"
    )
    
    GADMobileAds.sharedInstance().start(completionHandler: nil)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "loadNativeAd":
      loadNativeAd(call, result: result)
    case "disposeNativeAd":
      disposeNativeAd(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func loadNativeAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "Arguments must be a dictionary", detail: nil))
      return
    }

    let isTesting = args["isTesting"] as? Bool ?? false
    let adId = isTesting ? "ca-app-pub-3940256099942544/3986624511" : args["adId"] as? String

    guard let finalAdId = adId else {
      result(FlutterError(code: "MISSING_AD_ID", message: "Ad ID is required", detail: nil))
      return
    }

    let adsCount = args["adsCount"] as? Int ?? 1
    let requestId = UUID().uuidString
    loadResults[requestId] = (loaded: [], failed: 0, total: adsCount, result: result)

    let multipleAdsOptions = GADMultipleAdsOptions()
    multipleAdsOptions.numberOfAds = adsCount

    let window = UIApplication.shared.connectedScenes
      .filter { $0.activationState == .foregroundActive }
      .map { $0 as? UIWindowScene }
      .compactMap { $0 }
      .first?.windows
      .filter { $0.isKeyWindow }.first

    adLoader = GADAdLoader(
      adUnitID: finalAdId,
      rootViewController: window?.rootViewController,
      adTypes: [.native],
      options: [multipleAdsOptions]
    )
    adLoader?.delegate = self
    adLoader?.load(GADRequest())
  }

  public func adLoader(_ adLoader: GADAdLoader, didReceive nativeAd: GADNativeAd) {
    let adId = UUID().uuidString
    var adMap = [String: Any?]()
    
    adMap["id"] = adId
    adMap["headline"] = nativeAd.headline
    adMap["body"] = nativeAd.body
    adMap["advertiser"] = nativeAd.advertiser
    adMap["cta"] = nativeAd.callToAction
    adMap["starRating"] = nativeAd.starRating
    adMap["store"] = nativeAd.store
    adMap["price"] = nativeAd.price
    
    // Icon
    if let adIcon = nativeAd.icon {
        var iconMap = [String: Any?]()
        iconMap["url"] = adIcon.imageURL?.absoluteString
        if let img = adIcon.image {
            iconMap["width"] = Double(img.size.width)
            iconMap["height"] = Double(img.size.height)
        }
        iconMap["scale"] = Double(adIcon.scale)
        adMap["icon"] = iconMap
    }
    
    // Images
    if let images = nativeAd.images, !images.isEmpty {
        let imageListData = images.map { adImage -> [String: Any?] in
            var imageMap = [String: Any?]()
            imageMap["url"] = adImage.imageURL?.absoluteString
            if let img = adImage.image {
                imageMap["width"] = Double(img.size.width)
                imageMap["height"] = Double(img.size.height)
            }
            imageMap["scale"] = Double(adImage.scale)
            return imageMap
        }
        adMap["images"] = imageListData
        adMap["cover"] = imageListData.first?["url"] as? String
    } else {
        adMap["images"] = nil
        adMap["cover"] = nil
    }
    
    // Media Content (Aspect Ratio)
    adMap["aspectRatio"] = Double(nativeAd.mediaContent.aspectRatio)
    
    adMap["adChoicesUrl"] = "https://adssettings.google.com/whythisad"
    adMap["adChoicesText"] = nativeAd.adChoicesInfo?.text

    self.nativeAds[adId] = nativeAd
    nativeAd.delegate = self

    finalizeRequest(adLoader, adMap: adMap)
  }

  public func nativeAdDidRecordImpression(_ nativeAd: GADNativeAd) {
      if let adId = findAdId(for: nativeAd) {
          invokeOnMain("onAdImpression", arguments: adId)
      }
  }

  public func nativeAdDidRecordClick(_ nativeAd: GADNativeAd) {
      if let adId = findAdId(for: nativeAd) {
          invokeOnMain("onAdClicked", arguments: adId)
      }
  }

  public func nativeAdWillPresentScreen(_ nativeAd: GADNativeAd) {
      if let adId = findAdId(for: nativeAd) {
          invokeOnMain("onAdOpened", arguments: adId)
      }
  }

  public func nativeAdDidDismissScreen(_ nativeAd: GADNativeAd) {
      if let adId = findAdId(for: nativeAd) {
          invokeOnMain("onAdClosed", arguments: adId)
      }
  }

  private func findAdId(for nativeAd: GADNativeAd) -> String? {
      return nativeAds.first(where: { $0.value === nativeAd })?.key
  }

  private var channelInstance: FlutterMethodChannel?

  private func invokeOnMain(_ method: String, arguments: Any?) {
      DispatchQueue.main.async {
          self.channelInstance?.invokeMethod(method, arguments: arguments)
      }
  }

  private func finalizeRequest(_ adLoader: GADAdLoader, adMap: [String: Any?]?) {
      for (key, var val) in loadResults {
          if let map = adMap {
              val.loaded.append(map)
              loadResults[key] = val
          }
          if val.loaded.count + val.failed == val.total {
              val.result(val.loaded)
              loadResults.removeValue(forKey: key)
          }
      }
  }

  public func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: Error) {
    for (key, var val) in loadResults {
        val.failed += 1
        loadResults[key] = val
        if val.loaded.count + val.failed == val.total {
            if val.loaded.isEmpty {
                val.result(FlutterError(code: "LOAD_FAILED", message: error.localizedDescription, detail: nil))
            } else {
                val.result(val.loaded)
            }
            loadResults.removeValue(forKey: key)
        }
    }
  }


  private func disposeNativeAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let id = args["id"] as? String else {
      result(nil)
      return
    }

    self.nativeAds.removeValue(forKey: id)
    result(nil)
  }
}
